src/react/branching.js

```
@@ -39,78 +39,132 @@ export type BranchStatusEnum = "ROOT" | "NO_BRANCH" | "NEW_BRANCH" | "BRANCH";
-    let applyBranchedLogic = false;
+  // we first check our "parent" value, that was used to get the inlined value
+  const findMismatchingNonHostTypes = (x: Value, y: Value, arrayDepth: number): void => {
+    if (x instanceof ObjectValue && isReactElement(x) && y instanceof ObjectValue && isReactElement(y)) {
+      let xType = getProperty(realm, x, "type");
+      let yType = getProperty(realm, y, "type");
 
-    for (let i = 0; i < this._branchesToValidate.length; i++) {
-      let { type } = this._branchesToValidate[i];
-      if (reactElementType === undefined) {
-        reactElementType = type;
-      } else if (type !== reactElementType) {
-        // the types of the ReactElements do not match, so apply branch logic
-        applyBranchedLogic = true;
-        break;
-      }
-    }
-    if (applyBranchedLogic) {
-      for (let i = 0; i < this._branchesToValidate.length; i++) {
-        this._applyBranchedLogicValue(realm, reactSerializerState, this._branchesToValidate[i].value);
+      if (xType instanceof StringValue && yType instanceof StringValue) {
+        let xProps = getProperty(realm, x, "props");
+        let yProps = getProperty(realm, y, "props");
+        if (xProps instanceof ObjectValue && yProps instanceof ObjectValue) {
+          let xChildren = getProperty(realm, xProps, "children");
+          let yChildren = getProperty(realm, yProps, "children");
+
+          if (xChildren instanceof Value && yChildren instanceof Value) {
+            findMismatchingNonHostTypes(xChildren, yChildren, arrayDepth);
+          }
+        }
+      } else if (!xType.equals(yType)) {
+        return findMatchingComponentTypes(xType, yType);
       }
-      return true;
+    } else if (
+      ArrayValue.isIntrinsicAndHasWidenedNumericProperty(x) ||
+      ArrayValue.isIntrinsicAndHasWidenedNumericProperty(y)
+    ) {
+      // If either case is an unknown array, we do not know
+      // the contents of the array, so we cannot add keys
```

And is it safe to skip then? Why? Adding keys seems essential for correctness.

---

Out of scope for now. We definitely didn't support this before and this PR was more of a clean up than adding brand new functionality. We might be able to add some form of helper method at runtime that adds in keys – as the array is completely abstract to us.

---

Yeah, but it seems like in that case we should fail hard instead of emitting code that works differently.

---

Let's handle in a follow up still. Failing hard here may make sense once we have more tests around it. I've updated https://github.com/facebook/prepack/issues/1131 to include a point about them.